File: test-suite/src/test/java/io/micronaut/docs/inject/scope/RefreshEventSpec.java
package io.micronaut.docs.inject.scope;

import io.micronaut.context.ApplicationContext;
import io.micronaut.context.annotation.Requires;
import io.micronaut.context.env.Environment;
import io.micronaut.http.HttpRequest;
import io.micronaut.http.HttpResponse;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import io.micronaut.http.annotation.Post;
import io.micronaut.http.client.HttpClient;
import io.micronaut.runtime.context.scope.Refreshable;
import io.micronaut.runtime.context.scope.refresh.RefreshEvent;
import io.micronaut.runtime.server.EmbeddedServer;
import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;

import javax.annotation.PostConstruct;
import javax.inject.Inject;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.concurrent.atomic.AtomicReference;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;
import static org.junit.Assert.*;

public class RefreshEventSpec {

    private static EmbeddedServer embeddedServer;

    private static HttpClient client;


    @BeforeClass
    public static void setup(){
        embeddedServer = ApplicationContext.run(EmbeddedServer.class, new HashMap<String, Object>() {{
            put("spec.name", "RefreshEventSpec");
            put("spec.lang", "java");
        }}, Environment.TEST);
        client = HttpClient.create(embeddedServer.getURL());
    }
    @AfterClass
    public static void teardown(){
        if(client != null){
            client.close();
        }
        if(embeddedServer != null){
            embeddedServer.close();
        }
    }

    @Test
    public void publishingARefreshEventDestroysBeanWithRefreshableScope() {
        String firstResponse = fetchForecast();

        assertTrue(firstResponse.contains("{\"forecast\":\"Scattered Clouds"));

        String secondResponse = fetchForecast();

        assertEquals(firstResponse, secondResponse);

        String response = evictForecast();

        assertEquals(
// tag::evictResponse[]
                "{\"msg\":\"OK\"}"
// end::evictResponse[]
                , response);

        AtomicReference<String> thirdResponse = new AtomicReference<>(fetchForecast());
        await().atMost(5, SECONDS).until(() -> {
            if (!thirdResponse.get().equals(secondResponse)) {
                return true;
            }
            thirdResponse.set(fetchForecast());
            return false;
        });

        assertNotEquals(thirdResponse.get(), secondResponse);
        assertTrue(thirdResponse.get().contains("\"forecast\":\"Scattered Clouds"));
    }

    public String fetchForecast() {
        return client.toBlocking().retrieve(HttpRequest.GET("/weather/forecast"));
    }

    public String evictForecast() {
        return client.toBlocking().retrieve(HttpRequest.POST("/weather/evict", new LinkedHashMap()));
    }

    //tag::weatherService[]
    @Refreshable // <1>
    public static class WeatherService {
        private String forecast;

        @PostConstruct
        public void init() {
            forecast = "Scattered Clouds " + new SimpleDateFormat("dd/MMM/yy HH:mm:ss.SSS").format(new Date());// <2>
        }

        public String latestForecast() {
            return forecast;
        }
    }
    //end::weatherService[]

    @Requires(property = "spec.name", value = "RefreshEventSpec")
    @Requires(property = "spec.lang", value = "java")
    @Controller("/weather")
    public static class WeatherController {
        @Inject
        private WeatherService weatherService;
        @Inject
        private ApplicationContext applicationContext;

        @Get(value = "/forecast")
        public HttpResponse<Map<String, String>> index() {
            LinkedHashMap<String, String> map = new LinkedHashMap<String, String>(1);
            map.put("forecast", weatherService.latestForecast());
            return HttpResponse.ok(map);
        }

        @Post("/evict")
        public HttpResponse<Map<String, String>> evict() {
            //tag::publishEvent[]
            applicationContext.publishEvent(new RefreshEvent());
            //end::publishEvent[]
            LinkedHashMap<String, String> map = new LinkedHashMap<String, String>(1);
            map.put("msg", "OK");
            return HttpResponse.ok(map);
        }
    }
}

File: test-suite-scala/src/test/scala/io/micronaut/docs/inject/scope/RefreshEventSpec.scala
/*
 * Copyright 2017-2026 original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.micronaut.docs.inject.scope

import io.micronaut.context.ApplicationContext
import io.micronaut.context.annotation.Requires
import io.micronaut.context.env.Environment
import io.micronaut.http.HttpRequest
import io.micronaut.http.HttpResponse
import io.micronaut.http.annotation.Controller
import io.micronaut.http.annotation.Get
import io.micronaut.http.annotation.Post
import io.micronaut.http.client.HttpClient
import io.micronaut.runtime.context.scope.Refreshable
import io.micronaut.runtime.context.scope.refresh.RefreshEvent
import io.micronaut.runtime.server.EmbeddedServer
import jakarta.annotation.PostConstruct
import jakarta.inject.Inject
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

import java.text.SimpleDateFormat
import java.util.Date
import java.util.LinkedHashMap
import scala.compiletime.uninitialized
import scala.jdk.CollectionConverters.*

class RefreshEventSpec:

  @Test
  def publishingARefreshEventDestroysBeanWithRefreshableScope(): Unit =
    val embeddedServer = ApplicationContext.run(
      classOf[EmbeddedServer],
      Map[String, Object](
        "spec.name" -> "RefreshEventSpec",
        "spec.lang" -> "scala"
      ).asJava,
      Environment.TEST
    )
    val client = embeddedServer.getApplicationContext.createBean(classOf[HttpClient], embeddedServer.getURL)
    try
      val firstResponse = fetchForecast(client)
      assertTrue(firstResponse.contains("{\"forecast\":\"Scattered Clouds"))

      val secondResponse = fetchForecast(client)
      assertEquals(firstResponse, secondResponse)

      assertEquals("{\"msg\":\"OK\"}", evictForecast(client))

      val thirdResponse = waitForForecastChange(client, secondResponse)
      assertNotEquals(thirdResponse, secondResponse)
      assertTrue(thirdResponse.contains("\"forecast\":\"Scattered Clouds"))
    finally
      client.close()
      embeddedServer.close()

  private def fetchForecast(client: HttpClient): String =
    client.toBlocking.retrieve(HttpRequest.GET("/weather/forecast"))

  private def evictForecast(client: HttpClient): String =
    client.toBlocking.retrieve(HttpRequest.POST("/weather/evict", new LinkedHashMap[String, String]()))

  private def waitForForecastChange(client: HttpClient, previous: String): String =
    var response = fetchForecast(client)
    val deadline = System.nanoTime() + java.util.concurrent.TimeUnit.SECONDS.toNanos(5)
    while response == previous && System.nanoTime() < deadline do
      Thread.sleep(100)
      response = fetchForecast(client)
    response

// tag::weatherService[]
@Refreshable // <1>
class WeatherService:
  private var forecast: String = uninitialized

  @PostConstruct
  def init(): Unit =
    forecast = "Scattered Clouds " + SimpleDateFormat("dd/MMM/yy HH:mm:ss.SSS").format(Date()) // <2>

  def latestForecast(): String = forecast
// end::weatherService[]

@Requires(property = "spec.name", value = "RefreshEventSpec")
@Requires(property = "spec.lang", value = "scala")
@Controller("/weather")
class WeatherController(
    private val weatherService: WeatherService,
    private val applicationContext: ApplicationContext
):

  @Get("/forecast")
  def index(): HttpResponse[java.util.Map[String, String]] =
    val body = LinkedHashMap[String, String](1)
    body.put("forecast", weatherService.latestForecast())
    HttpResponse.ok(body)

  @Post("/evict")
  def evict(): HttpResponse[java.util.Map[String, String]] =
    // tag::publishEvent[]
    applicationContext.publishEvent(RefreshEvent())
    // end::publishEvent[]
    val body = LinkedHashMap[String, String](1)
    body.put("msg", "OK")
    HttpResponse.ok(body)
